R/model_ivpin.R

```diff
@@ -0,0 +1,905 @@
+    e_3 <- log(1 - alpha) + Vb * log(eps) + Vs * log(eps) - 2 * eps * t
+    e_max <- pmax(e_1, e_2, e_3)
+    
+    log_likelihood <- -sum(log(exp(e_1 - e_max) + exp(e_2 - e_max) + exp(e_3 - e_max)) + e_max)
+    return(log_likelihood)
+  }
+
+  # Compute iVPIN with a rolling window
+
+  start_index <- samplength + 1
+
+  "
+  Easley et al. (2012b) derive the VPIN estimator based on the argument of two moment conditions, 
+  E[|VτB - VτS|] ≈ αμ and E[|VτB + VτS|] = 2ε + αμ, from the Poisson processes. 
+  According to Ke et al. (2017), the two moment conditions should instead be expressed as
+  E[|VτB-VτS||tτ;θ]≈ αμtτ and E[|VτB + VτS|tτ; θ ] = (2ε + αμ)tτ,
+  "
+  total_arrival_rate <- (bucketdata$agg.bvol + bucketdata$agg.svol) / bucketdata$duration
+  informed_arrival <- abs(bucketdata$agg.bvol - bucketdata$agg.svol) / bucketdata$duration
+
+  t <- bucketdata$duration
+  Vb <- bucketdata$agg.bvol
+  Vs <- bucketdata$agg.svol
+
+  ivpin <- numeric(nrow(bucketdata))  # Ensure ivpin is initialized
+
+  for (i in start_index:nrow(bucketdata)) {
+    j <- i - samplength
+    parms <- rep(-Inf, 4)  # alpha, delta, mu, eps
+    
+    log_lik <- Inf
+    flag <- -Inf
+    
+    best_log_likelihood <- Inf
+    exit_flag <- FALSE
+    
+    if (i == start_index || i %% 10 == 0) {
+      for (alpha_init in seq(0.1, 0.9, by = 0.2)) {
+        for (delta_init in seq(0.1, 0.9, by = 0.2)) {
+          mu_init <- mean(informed_arrival[j:i]) / alpha_init
+          eps_init <- mean(abs(total_arrival_rate[j:i] - informed_arrival[j:i])) / 2
+          
+          initial_guess <- c(logit(max(min(alpha_init, 0.999), 0.001)), 
+                             logit(max(min(delta_init, 0.999), 0.001)), 
+                             sqrt(mu_init), 
+                             sqrt(eps_init))
+          tryCatch({
+            result <- optim(initial_guess, compute_log_likelihood, t = t[j:i], Vb = Vb[j:i], Vs = Vs[j:i], method = "BFGS")
+          }, error = function(e) {
+            conditionMessage(e)
+          })
+          
+          if (result$value < best_log_likelihood && is.finite(result$value)) {
+            best_params <- result$par
+            best_log_likelihood <- result$value
+            exit_flag <- TRUE
+          }
+        }
+      }
+    } else {
+      initial_guess <- c(logit(best_params[1]), logit(best_params[2]), sqrt(best_params[3]), sqrt(best_params[4]))
+      tryCatch({
+        result <- optim(initial_guess, compute_log_likelihood, t = t[j:i], Vb = Vb[j:i], Vs = Vs[j:i], method = "BFGS")
+      }, error = function(e) {
+        conditionMessage(e)
+      })
+      
+      if (result$value < best_log_likelihood && is.finite(result$value)) {
+        best_params <- result$par
+        exit_flag <- TRUE
+      } else {
+        for (alpha_init in seq(0.1, 0.9, by = 0.2)) {
+          for (delta_init in seq(0.1, 0.9, by = 0.2)) {
+            mu_init <- informed_arrival[i] / alpha_init
```

Comment:
@alecriste I think I see the logic behind your work. Isn't there a redundancy here in the code? It is exactly the same code as above. Wouldn't be better to have a function that we can call whenever we need it?